### ISSUE TITLE
self._parser is None during _pre_argument_parsing 

### ISSUE THREAD

**Issue by ghost** [datafolklabs/cement#569]
Defining _pre_argument_parsing on a second controller class isn't working but for some reason it does work just fine on the other class. I've tested the order the controllers are listed as well as having the function on both of the classes and only 1 of them. Every time, Base2 crashes.

**System Information**

- Cement Version: 3.0.2
- Python Version: 3.6.8
- Operating System and Version: Linux-4.18.0-20-generic-x86_64-with-debian-buster-sid

**Steps to Reproduce (Bugs Only)**

1. add function:
    ```python
    def _pre_argument_parsing(self):
        p = self._parser
        p.add_argument('-b', '--bar',
                       help='my bar option',
                       dest='bar')
        pass
   ```
2. `$ cara -h`
3. Crash
```
  File "/home/stelios/Documents/Projects/Cara/env/bin/cara", line 11, in <module>
    load_entry_point('cara', 'console_scripts', 'cara')()
  File "/home/stelios/Documents/Projects/Cara/cara/main.py", line 67, in main
    app.run()
  File "/home/stelios/Documents/Projects/Cara/env/lib/python3.6/site-packages/cement/core/foundation.py", line 916, in run
    return_val = self.controller._dispatch()
  File "/home/stelios/Documents/Projects/Cara/env/lib/python3.6/site-packages/cement/ext/ext_argparse.py", line 771, in _dispatch
    contr._pre_argument_parsing()
  File "/home/stelios/Documents/Projects/Cara/cara/controllers/base.py", line 77, in _pre_argument_parsing
    p.add_argument('-b', '--bar',
AttributeError: 'NoneType' object has no attribute 'add_argument'

```
**Example**

The following is an [MCVE] (*Minimal Complete
Verifiable Example*) that reproduces the issue:

## app.py
```python

class Cara(App):
    """Cara primary application."""

    class Meta:
        label = 'cara'

        # configuration defaults
        config_defaults = CONFIG

        # call sys.exit() on close
        close_on_exit = True

        # load additional framework extensions
        extensions = [
            'yaml',
            'colorlog',
            'jinja2',
        ]

        # configuration handler
        config_handler = 'yaml'

        # configuration file suffix
        config_file_suffix = '.yml'

        # set the log handler
        log_handler = 'colorlog'

        # set the output handler
        output_handler = 'jinja2'

        # register handlers
        handlers = [
            Base,
            Base2
        ]

        hooks = [
            ('post_setup', database.extend_sqlite)
        ]
```

## cara/controllers/base.py
```python3


from cement import Controller, ex
from . import VERSION_BANNER


class Base(Controller):
    class Meta:
        label = 'base'

        # text displayed at the top of --help output
        description = 'Cara malware classifier'

        # text displayed at the bottom of --help output
        epilog = 'Usage: cara query file [file,...]'

        # controller level arguments. ex: 'cara --version'
        arguments = [

            (['-v', '--version'],
             {'action': 'version',
              'version': VERSION_BANNER}),
        ]

    def _pre_argument_parsing(self):
        p = self._parser
        p.add_argument('-b', '--bar',
                       help='my bar option',
                       dest='bar')
        pass

    def _default(self):
        """Default action if no sub-command is passed."""

        self.app.args.print_help()

    @ex(
        help='example sub command1',

        # sub-command level arguments. ex: 'cara command1 --foo bar'
        arguments=[
            # add a sample foo option under subcommand namespace
            (['-f', '--foo'],
             {'help': 'notorious foo option',
              'action': 'store',
              'dest': 'foo'}),
        ],
    )
    def command1(self):
        """Example sub-command."""

        data = {
            'foo': 'bar',
        }

        ### do something with arguments
        if self.app.pargs.foo is not None:
            data['foo'] = self.app.pargs.foo

        self.app.render(data, 'terminal_query.jinja2')


class Base2(Controller):
    class Meta:
        label = 'base2'

        # text displayed at the top of --help output
        description = 'Cara malware classifier'

        # text displayed at the bottom of --help output
        epilog = 'Usage: cara query file [file,...]'

        # controller level arguments. ex: 'cara --version'
        arguments = [
        ]

    def _pre_argument_parsing(self):
        p = self._parser
        p.add_argument('-b', '--bar',
                       help='my bar option',
                       dest='bar')
        pass

    def _default(self):
        """Default action if no sub-command is passed."""

        self.app.args.print_help()

    @ex(
        help='example sub command1',

        # sub-command level arguments. ex: 'cara command1 --foo bar'
        arguments=[
            # add a sample foo option under subcommand namespace
            (['-f', '--foo'],
             {'help': 'notorious foo option',
              'action': 'store',
              'dest': 'foo'}),
        ],
    )
    def command3(self):
        """Example sub-command."""

        data = {
            'foo': 'bar',
        }

        ### do something with arguments
        if self.app.pargs.foo is not None:
            data['foo'] = self.app.pargs.foo

        self.app.render(data, 'terminal_query.jinja2')
```

**Comment by derks** [comment]
@namedLambda Thank you for the detailed report. This appears to be a bug (or undocumented limitation) for `embedded` controllers.

The default `stacked_type = 'embedded'` ... if you change it to `nested` and test, it works.  I will need to review the argparse extension and pick this apart.

Question: Is the use of `_pre_argument_parsing` necessary here, or could you maybe put the arguments in `Base2.Meta.arguments`? (I assume maybe this was just a minimal example to demonstrate the issue, but just want to verify).

Thanks.

---

**Example:**

```
$ python3 myscript.py
Inside Third Controller: ArgparseArgumentHandler(prog='myscript third', usage=None, description=None, formatter_class=<class 'argparse.RawDescriptionHelpFormatter'>, conflict_handler='error', add_help=True)
Inside Second Controller: None
INFO: Testing 1, 2, 3
```

**myscript.py**:

```python
from cement import App, CaughtSignal, Controller, ex, get_version

VERSION = (0, 0, 1, 'alpha', 0)

VERSION_BANNER = """
myscript v%s
""" % get_version()


class Base(Controller):
    class Meta:
        label = 'base'

        arguments = [
            ### add a version banner
            ( [ '-v', '--version' ],
              { 'action'  : 'version',
                'version' : VERSION_BANNER } ),
        ]


    def _default(self):
        """Default action if no sub-command is passed."""

        self.app.log.info('Testing 1, 2, 3')


    @ex(
        help='example sub command1',
        arguments=[
            ### add a sample foo option under subcommand namespace
            ( [ '-f', '--foo' ],
              { 'help' : 'notorious foo option',
                'action'  : 'store',
                'dest' : 'foo' } ),
        ],
    )
    def command1(self):
        """Example sub-command."""

        print('Inside Base.command1')

        ### do something with arguments
        if self.app.pargs.foo is not None:
            print('Foo Argument > %s' % self.app.pargs.foo)


class Second(Controller):
    class Meta:
        label = 'second'
        stacked_on = 'base'
        stacked_type = 'embedded'

    def _pre_argument_parsing(self):
        print(f'Inside Second Controller: {self._parser}')

class Third(Controller):
    class Meta:
        label = 'third'
        stacked_on = 'base'
        stacked_type = 'nested'

    def _pre_argument_parsing(self):
        print(f'Inside Third Controller: {self._parser}')


class MyApp(App):

    class Meta:
        # application label
        label = 'myscript'
        extensions = ['colorlog']
        log_handler = 'colorlog'

        # register handlers
        handlers = [
            Base,
            Second,
            Third,
        ]

        # call sys.exit() on close
        close_on_exit = True


def main():
    with MyApp() as app:
        try:
            app.run()
        except CaughtSignal as e:
            # Default Cement signals are SIGINT and SIGTERM, exit 0 (non-error)
            print('\n%s' % e)
            app.exit_code = 0


if __name__ == '__main__':
    main()
```

**Comment by pwhipp** [comment]
I have the same problem.
Adding the arguments to the meta would be great but I'm adding an exclusive group of options which in argparse I'd do thus:
```

    action = parser.add_mutually_exclusive_group(required=True)
    action.add_argument('--tic', action='store_true')
    action.add_argument('--top', action='store_true')
    action.add_argument('--toe', action='store_true')
```
In cement I believe I need to get hold of the relevant sub parser to do this (I don't want to add the group at the top level so I can't use app.args)

**Comment by derks** [comment]
@pwhipp apologies that this took forever... this is in `master`, and will be released in the next stable/3.0.6.
